Add CONTRIBUTING.md (#242)
radicle-dev/radicle-registry: CONTRIBUTING.md

# Abbreviations

> The usage of common abbreviations in the codebase is appreciated as long as they:

Reviewer: Picking up on my [previous comment](https://github.com/radicle-dev/radicle-registry/pull/242#discussion_r389567057). To bring some more clarity around this I’d suggest to use the standard [RFC terminology](https://github.com/radicle-dev/radicle-registry/pull/242#discussion_r389567057).

```suggestion
You may use common abbreviations in the codebase as long as they:
```

Author: Fixed

- don't obscure their meaning for the sole sake of saving a few characters
- are common programming abbreviations (e.g. `num`, `idx` or `iter`)

The usage of domain-specific abbreviations is generally strongly discouraged.

> The only exception, when using abbreviations is mandatory, is for the names on this list:

Reviewer: Continuing the suggestion from above and being more prescriptive
```suggestion
For consistency we always use the following domain-specific abbreviation. You
may not use any other domain-specific abbreviations.
```

Author: Fixed

> # Use of abbreviations

Reviewer: How do you feel about moving this section to the bottom? I think it’s the least relevant for users looking at this document for the first time.

Author: I've put it on the top to make it the most accessible as a quick lookup and most visible for newcomers, because it's not a common thing for a project to have a dictionary.

Reviewer: That reasoning makes sense, too. Let’s keep it as it is.

> Usage of abbreviations in the codebase is highly recommended for brevity.

Reviewer: When I read this the use of abbreviations sounds optional to me. This leaves the door open for inconsistencies like using `Tx` in some instances and `Transaction` in others. @NunoAlexandre’s suggestion has the same issue:

> The usage of abbreviations in the codebase is appreciated as long as they

If we want to go with abbreviations we should make use of them mandatory.

Reviewer 2: @geigerzaehler the argument is not to _sometimes_ use `tx` and other times use `transaction`. It's more to say `p` is not enough, `proj` is, just an example.

Reviewer: I understand that this is the intention. But I could be misread to mean the former so I think we should clarify it.

Reviewer 2: Ah, I see what you mean 👍

Some abbreviations are widely understood and used in the software developer community in general,
e.g. `num`, `idx` or `iter`. They should be used whenever they don't obscure the meaning.

Other abbreviations are more domain-specific for blockchains, Substrate or just this project. They
should be used whenever they don't obscure the meaning as well, but only if they are listed here:

- Tx - transaction
- PoW - proof of work

> - Org - organization

Author: This is a neat improvement! I'm not sure about the `are well-established and agreed on in the context of the Blockchain, Substrate, or this project itself` part. This sentence removes the need for the list below, all of these 3 abbreviations are well established and agreed on. This is fine with me, but you were the first one to notice and rise alarm that they may not be understood by random contributors.

> - Org - organization

Reviewer 2: Yeah, that's a good point.  We can just have the list to be explicit about which of those common abbreviations we use and agree on?

> - Org - organization

Author: So we revert to not allowing well-established blockchain-specific abbreviations and meintening the list?

> - Org - organization

Reviewer 2: yeah ok, whitelisting sounds like a good idea. I don't think there will be a lot of friction but yeah, whitelisting seems better here. 👍

# Git Flow

TL;DR: This repository follows the Git rebase flow for the most part. Each feature branch must
be reviewed and approved in a pull request. Once ready to merge, consider squashing the branch's
commits when the separate commits don't add value, rebase it, force push with lease, and merge it
via the GitHub UI.

> ## **Branches**

Reviewer: We do we mark this as bold?

Reviewer 2: No need

2. Do your magic
3. Keep your branch up to date by rebasing it from its base branch
4. Delete the branch after its been both approved and merged. GitHub does this automatically.

## **Commits**

1. Make sure you author your commits with the right username and email
2. Follow the git commit convention:
  - Use the imperative mood in the subject line
  - Limit the subject line to 50 chars
  - Capitalise the subject line
  - Wrap the description at 72 characters
  - Have the description preferably explaining what and why instead of how
  - Separate the subject from the body with an empty line

# Documentation

- The code must be documented using Rustdoc comments.
- Strive for self-documenting code.
- The reader of the documentation is expected to have general knowledge in the fields of blockchain
  and cryptocurrency.
- Leave additional comments explaining 'why' rather than 'how', striving to have the code clean
  and elegant to make this aim viable.
- Document all the public items of a module especially well given that they constitute its API.
- Document private items when the reader's understanding to fully grasp its usage or implementation

>   is insufficient, despite the limited context of such items.

Reviewer: I’d prefer to move these to the top since they are the most important points. They tell you the “what”, the rest the “how”.
